src/test/java/com/google/firebase/FirebaseAppTest.java

```diff
@@ -345,6 +350,99 @@ public void testProactiveTokenRefresh() throws Exception {
+    Assert.assertNotNull(tokenRefresher);
+
+    // Get the initial token
+    TestOnlyImplFirebaseTrampolines.getToken(firebaseApp, true);
+
+    CredentialsChangedListener listener = mock(CredentialsChangedListener.class);
+    ImplFirebaseTrampolines.getCredentials(firebaseApp).addChangeListener(listener);
+
+    firebaseApp.startTokenRefresher();
+
+    // Since there is already a valid token, which won't expire for another hour, the refresher
+    // should not refresh the credential at this point in time.
+    tokenRefresher.simulateDelay(0);
+    verify(listener, never()).onChanged(Mockito.any(OAuth2Credentials.class));
+
+    tokenRefresher.simulateDelay(55);
+    verify(listener, times(1)).onChanged(Mockito.any(OAuth2Credentials.class));
+
+    tokenRefresher.simulateDelay(20);
+    verify(listener, times(1)).onChanged(Mockito.any(OAuth2Credentials.class));
+
+    tokenRefresher.simulateDelay(35);
+    verify(listener, times(2)).onChanged(Mockito.any(OAuth2Credentials.class));
+  }
+
+  @Test
+  public void testTokenRefresherStateMachine() {
+    FirebaseApp firebaseApp = FirebaseApp.initializeApp(getMockCredentialOptions(), "myApp");
+    final AtomicInteger scheduleCalls = new AtomicInteger(0);
+    final AtomicInteger cancelCalls = new AtomicInteger(0);
+    TokenRefresher refresher = new TokenRefresher(firebaseApp) {
+      @Override
+      protected void cancelPrevious() {
+        cancelCalls.incrementAndGet();
+      }
+
+      @Override
+      protected void scheduleNext(Callable<Void> task, long delayMillis) {
+        scheduleCalls.incrementAndGet();
+      }
+    };
+    assertEquals(0, scheduleCalls.get());
+
+    // Both schedule and cancel should be called.
+    // (we call cancel before each new scheduling attempt)
+    refresher.start();
+    assertEquals(1, scheduleCalls.get());
+    assertEquals(1, cancelCalls.get());
+
+    // Shouldn't change state
+    refresher.start();
+    assertEquals(1, scheduleCalls.get());
+    assertEquals(1, cancelCalls.get());
+
+    refresher.stop();
+    assertEquals(1, scheduleCalls.get());
+    assertEquals(2, cancelCalls.get());
+
+    // Shouldn't change state
+    refresher.stop();
+    assertEquals(1, scheduleCalls.get());
+    assertEquals(2, cancelCalls.get());
+  }
+
+  @Test
+  public void testTokenRefresherStateMachineEarlyTermination() {
+    FirebaseApp firebaseApp = FirebaseApp.initializeApp(getMockCredentialOptions(), "myApp");
+    final AtomicInteger scheduleCalls = new AtomicInteger(0);
+    final AtomicInteger cancelCalls = new AtomicInteger(0);
+    TokenRefresher refresher = new TokenRefresher(firebaseApp) {
+      @Override
+      protected void cancelPrevious() {
+        cancelCalls.incrementAndGet();
+      }
+
+      @Override
+      protected void scheduleNext(Callable<Void> task, long delayMillis) {
+        scheduleCalls.incrementAndGet();
+      }
+    };
+    // stop() is allowed here, but since we didn't start(), no measurable state change
```

Comment:
Nit: I know there are good reasons to keep this as a separate test case, but I would personally combine it with the previous test.

Reply:
Done